Template specializations

**Summary**

Wrappers are created for methods in template specializations that may not have been implemented. In the example below, the `Foo(unsigned u)` constructor is not implemented in the specialization.

Foo.hpp
```c++
template<unsigned DIM_A, unsigned DIM_B>
class Foo : public AbstractFoo<DIM_A, DIM_B>
{
public:
  Foo(unsigned u);
  Foo(unsigned u, double d);
};

// Specialization
template<unsigned DIM_B>
class Foo<1, DIM_B> : public AbstractFoo<1,DIM_B>
{
public:
  Foo(unsigned u, double d);
  Foo(unsigned u);
};
```

Foo.cpp
```c++
template<unsigned DIM_A, unsigned DIM_B>
Foo<DIM_A, DIM_B>::Foo(unsigned u) : AbstractFoo<DIM_A, DIM_B>(u)
{
  // implementation
}

template<unsigned DIM_A, unsigned DIM_B>
Foo<DIM_A, DIM_B>::Foo(unsigned u, double d) : AbstractFoo<DIM_A, DIM_B>(u, d)
{
  // implementation
}

// Specialization
template<unsigned DIM_B>
Foo<1, DIM_B>::Foo(unsigned u, double d) : AbstractFoo<1, DIM_B>(u, d)
{
  // implementation
}
```

The wrapper below is generated for `Foo<1,2>`:

```c++
py::class_<Foo1_2, Foo1_2_Overloads>, AbstractFoo<1,2> >(m, "Foo1_2")
  .def(py::init<unsigned int>(), py::arg("u"))
  .def(py::init<unsigned int, double>(), py::arg("u"), py::arg("d"))
  ```

This results in a linker error:

```
undefined reference to `Foo<1u, 2u>::Foo(unsigned int)'
```

--- a/cppwg/input/base_info.py
+++ b/cppwg/input/base_info.py
@@ -38,12 +38,14 @@ class BaseInfo:
         Do not include these methods.
     excluded_variables : List[str]
         Do not include these variables.
+    arg_type_excludes : List[str]
+        List of exclude patterns for arg types in methods.
     constructor_arg_type_excludes : List[str]
-        List of exclude patterns for ctors.
+        List of exclude patterns for arg types in constructors.
+    constructor_signature_excludes : List[List[str]]
+        List of exclude patterns for constructor signatures.
     return_type_excludes : List[str]
         List of exclude patterns for return types.
-    arg_type_excludes : List[str]
-        List of exclude patterns for arg types.
     name_replacements : Dict[str, str]
         A dictionary of name replacements e.g. {"double":"Double", "unsigned
         int":"Unsigned"}
@@ -62,9 +64,10 @@ class BaseInfo:
         self.custom_generator: Optional[str] = None
         self.excluded_methods: List[str] = []
         self.excluded_variables: List[str] = []
-        self.constructor_arg_type_excludes: List[str] = []
-        self.return_type_excludes: List[str] = []
         self.arg_type_excludes: List[str] = []
+        self.constructor_arg_type_excludes: List[str] = []
+        self.constructor_signature_excludes: List[List[str]] = []
+        self.return_type_excludes: List[str] = []
         self.name_replacements: Dict[str, str] = {
             "double": "Double",
             "unsigned int": "Unsigned",

--- a/cppwg/parsers/package_info_parser.py
+++ b/cppwg/parsers/package_info_parser.py
@@ -123,6 +123,7 @@ class PackageInfoParser:
             "pointer_call_policy": None,
             "reference_call_policy": None,
             "constructor_arg_type_excludes": None,
+            "constructor_signature_excludes": None,
             "excluded_methods": [],
             "excluded_variables": [],
             "custom_generator": None,

--- a/cppwg/writers/constructor_writer.py
+++ b/cppwg/writers/constructor_writer.py
@@ -94,35 +94,51 @@ class CppConstructorWrapperWriter(CppBaseWrapperWriter):
         ):
             return True
 
-        # Check for excluded argument patterns
+        # Get arg type strings with spaces removed
+        # e.g. ::std::vector<unsigned int> const & -> ::std::vector<unsignedint>const&
+        arg_types = [
+            x.decl_string.replace(" ", "") for x in self.ctor_decl.argument_types
+        ]
+
+        # Exclude constructors with "iterator" in args
+        for arg_type in arg_types:
+            if "iterator" in arg_type.lower():
+                return True
+
+        # Exclude constructors with args matching patterns in calldef_excludes
         calldef_excludes = [
             x.replace(" ", "")
             for x in self.class_info.hierarchy_attribute_gather("calldef_excludes")
         ]
+        for arg_type in arg_types:
+            if arg_type in calldef_excludes:
+                return True
 
+        # Exclude constructors with args matching patterns in constructor_arg_type_excludes
         ctor_arg_type_excludes = [
             x.replace(" ", "")
             for x in self.class_info.hierarchy_attribute_gather(
                 "constructor_arg_type_excludes"
             )
         ]
-
-        for arg_type in self.ctor_decl.argument_types:
-            # e.g. ::std::vector<unsigned int> const & -> ::std::vector<unsignedint>const&
-            arg_type_str = arg_type.decl_string.replace(" ", "")
-
-            # Exclude constructors with "iterator" in args
-            if "iterator" in arg_type_str.lower():
+        for exclude_type in ctor_arg_type_excludes:
+            if exclude_type in arg_type:
                 return True
 
-            # Exclude constructors with args matching calldef_excludes
-            if arg_type_str in calldef_excludes:
-                return True
+        # Exclude constructors matching a signature in constructor_signature_excludes
+        ctor_signature_excludes = self.class_info.hierarchy_attribute_gather(
+            "constructor_signature_excludes"
+        )
 
-            # Exclude constructurs with args matching constructor_arg_type_excludes
-            for excluded_type in ctor_arg_type_excludes:
-                if excluded_type in arg_type_str:
-                    return True
+        for exclude_types in ctor_signature_excludes:
+            if len(exclude_types) != len(arg_types):
+                continue
+
+            if all(
+                exclude_type in arg_type
+                for arg_type, exclude_type in zip(arg_types, exclude_types)
+            ):
+                return True
 
         return False
 

--- a/cppwg/writers/constructor_writer.py
+++ b/cppwg/writers/constructor_writer.py
@@ -122,8 +122,9 @@ class CppConstructorWrapperWriter(CppBaseWrapperWriter):
             )
         ]
         for exclude_type in ctor_arg_type_excludes:
-            if exclude_type in arg_type:
-                return True
+            for arg_type in arg_types:
+                if exclude_type in arg_type:
+                    return True
 
         # Exclude constructors matching a signature in constructor_signature_excludes
         ctor_signature_excludes = self.class_info.hierarchy_attribute_gather(

--- a/examples/shapes/src/geometry/Point.cpp
+++ b/examples/shapes/src/geometry/Point.cpp
@@ -19,6 +19,14 @@ Point<DIM>::Point(double x, double y, double z)
 }
 
 template <unsigned DIM>
+Point<DIM>::Point(int x, int y, int z)
+    : Point(static_cast<double>(x),
+            static_cast<double>(y),
+            static_cast<double>(z))
+{
+}
+
+template <unsigned DIM>
 Point<DIM>::~Point()
 {
 }

--- a/examples/shapes/src/geometry/Point.hpp
+++ b/examples/shapes/src/geometry/Point.hpp
@@ -32,6 +32,11 @@ public:
     Point(double x, double y, double z = (DIM - DIM));
 
     /**
+     * Constructor to exclude from wrapping
+     */
+    Point(int x, int y, int z = 0);
+
+    /**
      * Destructor
      */
     ~Point();

--- a/examples/shapes/wrapper/package_info.yaml
+++ b/examples/shapes/wrapper/package_info.yaml
@@ -16,14 +16,6 @@ source_includes:
 # Exclude default arguments from wrapped methods.
 exclude_default_args: False
 
-# Exclude any methods that have these arg types.
-# arg_type_excludes:
-#   - double
-
-# Exclude any constructors that have these arg types.
-# constructor_arg_type_excludes:
-#   - double
-
 # Signature/replacement settings for explicit template instantiations.
 template_substitutions:
   - signature: <unsigned DIM>
@@ -52,6 +44,19 @@ modules:
         # List of methods that should not be wrapped.
         excluded_methods:
           - ExcludedMethod
+
+        # Exclude any methods that have these arg types.
+        # arg_type_excludes:
+        #   - double
+
+        # Exclude any constructors that have these arg types.
+        # constructor_arg_type_excludes:
+        #   - double
+
+        # Exclude any constructors that have these signatures.
+        # constructor_signature_excludes:
+        #   - [double]
+        #   - [double, int]
 
         # Path to a custom script for generating wrappers for this class.
         # CPPWG_SOURCEROOT points to the supplied source root directory.

--- a/examples/shapes/wrapper/package_info.yaml
+++ b/examples/shapes/wrapper/package_info.yaml
@@ -54,9 +54,8 @@ modules:
         #   - double
 
         # Exclude any constructors that have these signatures.
-        # constructor_signature_excludes:
-        #   - [double]
-        #   - [double, int]
+        constructor_signature_excludes:
+          - [int, int, int]
 
         # Path to a custom script for generating wrappers for this class.
         # CPPWG_SOURCEROOT points to the supplied source root directory.

